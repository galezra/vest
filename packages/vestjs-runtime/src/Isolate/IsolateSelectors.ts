import { Maybe } from 'vest-utils';

import { TIsolate } from 'Isolate';
import { IsolateKeys } from 'IsolateKeys';

export function isIsolateType<I extends TIsolate>(
  node: Maybe<TIsolate>,
  type: string,
): node is I {
  return node?.[IsolateKeys.Type] === type;
}

export function isSameIsolateType<A extends TIsolate, B extends TIsolate>(
  a: A,
  b: B,
): boolean {
  return isIsolateType(a, b[IsolateKeys.Type]);
}

export function isSameIsolateIdentity<A extends TIsolate, B extends TIsolate>(
  a: A,
  b: B,
): boolean {
  return Object.is(a, b) || (isSameIsolateType(a, b) && a.key === b.key);
}
