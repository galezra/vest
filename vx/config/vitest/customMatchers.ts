import { isFunction } from 'vest-utils';

expect.extend({
  isDeepCopyOf,
});

// eslint-disable-next-line complexity
function isDeepCopyOf(
  source: any,
  clone: any,
): { pass: boolean; message: () => string } {
  const queue = [[source, clone]];

  outer: while (queue.length) {
    const [source, clone] = (queue as any[]).shift();

    if (!source || typeof source !== 'object') {
      if (!isFunction(clone)) {
        if (clone !== source) {
          return {
            pass: false,
            message: () => `Source and clone are not identical`,
          };
        }
      }
      continue;
    }

    if (clone === source) {
      return {
        pass: false,
        message: () =>
          `Source and clone are the same object. Expected a deep copy. ${JSON.stringify(
            source,
          )}===${JSON.stringify(clone)}`,
      };
    }

    // Short circuit
    if (
      (clone && !source) ||
      (source && !clone) ||
      typeof source !== typeof clone
    ) {
      return {
        pass: false,
        message: () =>
          `Source and clone are not of the same type: ${JSON.stringify(
            source,
          )} does not equal ${JSON.stringify(clone)}`,
      };
    }

    if (Array.isArray(source)) {
      // Short circuit
      if (!Array.isArray(clone) || source.length !== clone.length) {
        return {
          pass: false,
          message: () =>
            `source and clone arrays are not identical. ${JSON.stringify(
              source,
            )} does not equal ${JSON.stringify(clone)}`,
        };
      }
      source.forEach((_, i) => {
        queue.push([source[i], clone[i]]);
      });
    } else if (typeof source === 'object') {
      Object.keys(source).forEach(key => queue.push([source[key], clone[key]]));
    }

    continue outer;
  }

  return { pass: true, message: () => 'success' };
}
