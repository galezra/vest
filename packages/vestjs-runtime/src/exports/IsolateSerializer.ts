import { ErrorStrings } from 'ErrorStrings';
import {
  Nullable,
  hasOwnProperty,
  invariant,
  isNullish,
  isStringValue,
  text,
} from 'vest-utils';
import { expandObject, minifyObject } from 'vest-utils/minifyObject';

import { TIsolate } from 'Isolate';
import { ExcludedFromDump, IsolateKeys } from 'IsolateKeys';
import { IsolateMutator } from 'IsolateMutator';

export class IsolateSerializer {
  // eslint-disable-next-line max-statements, complexity, max-lines-per-function
  static deserialize(node: Record<string, any> | TIsolate | string): TIsolate {
    // Validate the root object
    const root = (
      isStringValue(node) ? JSON.parse(node) : ({ ...node } as TIsolate)
    ) as [any, any];

    const expanded = expandObject(...root);

    IsolateSerializer.validateIsolate(expanded);

    const queue = [expanded];

    // Iterate over the queue until it's empty
    while (queue.length) {
      // Get the next item from the queue
      const current = queue.shift();

      if (!current) {
        continue;
      }

      const children = current.children;

      // If there are no children, nothing to do.
      if (!children) {
        continue;
      }

      // Copy the children and set their parent to the current node.
      current.children = children.map(child => {
        const nextChild = { ...child };

        IsolateMutator.setParent(nextChild, current);
        queue.push(nextChild);

        // If the child has a key, add it to the parent's keys.
        const key = nextChild.key;

        if (key) {
          current.keys = current.keys ?? {};
          current.keys[key] = nextChild;
        }

        return nextChild;
      });
    }

    return expanded;
  }

  static serialize(
    isolate: Nullable<TIsolate>,
    replacer: (value: any, key: string) => any,
  ): string {
    if (isNullish(isolate)) {
      return '';
    }

    const minified = minifyObject(isolate, (value: any, key: string) => {
      if (ExcludedFromDump.has(key as any)) {
        return undefined;
      }
      return replacer(value, key);
    });

    return JSON.stringify(minified);
  }

  static validateIsolate(
    node: Record<string, any> | TIsolate,
  ): asserts node is TIsolate {
    invariant(
      hasOwnProperty(node, IsolateKeys.Type),
      text(ErrorStrings.INVALID_ISOLATE_CANNOT_PARSE),
    );
  }
}
