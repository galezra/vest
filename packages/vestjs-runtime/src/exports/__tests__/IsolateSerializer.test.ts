import { Isolate, TIsolate } from 'Isolate';
import { IsolateSerializer } from 'IsolateSerializer';
import { Maybe } from 'vest-utils';
import { describe, it, expect, test } from 'vitest';

import { VestRuntime } from 'vestjs-runtime';

describe('IsolateSerializer', () => {
  describe('serialize', () => {
    it('Should produce serialized dump', () => {
      const { serialized } = createRoot();

      expect(serialized).toMatchSnapshot();
    });
  });

  describe('deserialize', () => {
    it('Should fully inflate the tree', () => {
      const { root, serialized } = createRoot();

      const inflated = IsolateSerializer.deserialize(serialized);

      expect(inflated).toMatchInlineSnapshot(`
        {
          "$type": "URoot",
          "children": [
            {
              "$type": "UChild_1",
              "data": {
                "some_data": true,
              },
              "parent": [Circular],
            },
            {
              "$type": "UChild_2",
              "parent": [Circular],
            },
            {
              "$type": "UChild_3",
              "parent": [Circular],
            },
          ],
          "data": {
            "some_data": true,
          },
        }
      `);
    });

    test('When data is empty, should not add data property', () => {
      const { serialized } = createRoot();
      const inflated = IsolateSerializer.deserialize(serialized);

      expect(inflated?.children?.[0]).toHaveProperty('data');
      expect(inflated?.children?.[1]).not.toHaveProperty('data');
      expect(inflated?.children?.[2]).not.toHaveProperty('data');
    });
  });

  describe('Custom Data Serialization', () => {
    it('Should serialize data with custom keys', () => {
      const { serialized } = createRoot({
        keys: {
          data: {
            some_data: 'sd',
          },
        },
      });

      const parsed = JSON.parse(serialized);
      expect(serialized).toMatchSnapshot();
    });

    describe('value serialization', () => {
      it('Should correctly expand values', () => {
        const { root } = createRoot();

        root.status = 'pending';
        // @ts-ignore
        root.children[0].status = 'done';
        // @ts-ignore
        root.children[1].status = 'failed';

        const minimap = {
          values: {
            status: {
              pending: 'p',
              done: 'd',
              failed: 'f',
            },
            $type: {
              URoot: 'UR',
              UChild_1: 'UC1',
              UChild_2: 'UC2',
              UChild_3: 'UC3',
            },
          },
        };

        const serialized = IsolateSerializer.serialize(root, minimap);
        const inflated = IsolateSerializer.deserialize(serialized, minimap);

        expect(inflated.status).toBe('pending');
        // @ts-ignore
        expect(inflated.children[0].status).toBe('done');
        // @ts-ignore
        expect(inflated.children[1].status).toBe('failed');
        expect(inflated).toMatchInlineSnapshot(`
          {
            "$type": "URoot",
            "children": [
              {
                "$type": "UChild_1",
                "data": {
                  "some_data": true,
                },
                "parent": [Circular],
                "status": "done",
              },
              {
                "$type": "UChild_2",
                "parent": [Circular],
                "status": "failed",
              },
              {
                "$type": "UChild_3",
                "parent": [Circular],
              },
            ],
            "data": {
              "some_data": true,
            },
            "status": "pending",
          }
        `);
      });
    });

    it('Should inflate with correct keys', () => {
      const { serialized } = createRoot({
        keys: {
          data: {
            some_data: 'sd',
          },
        },
      });

      const inflated = IsolateSerializer.deserialize(serialized, {
        keys: {
          data: {
            some_data: 'sd',
          },
        },
      });

      expect(inflated.data.some_data).toBe(true);
      expect(inflated).not.toHaveProperty('sd');
      expect(inflated).toEqual(
        IsolateSerializer.deserialize(createRoot().serialized),
      );
      expect(inflated).toMatchInlineSnapshot(`
        {
          "$type": "URoot",
          "children": [
            {
              "$type": "UChild_1",
              "data": {
                "some_data": true,
              },
              "parent": [Circular],
            },
            {
              "$type": "UChild_2",
              "parent": [Circular],
            },
            {
              "$type": "UChild_3",
              "parent": [Circular],
            },
          ],
          "data": {
            "some_data": true,
          },
        }
      `);
    });
  });
});

function withRunTime<T>(fn: CB<T>) {
  return VestRuntime.Run(VestRuntime.createRef({}), () => {
    return fn();
  });
}

function createRoot(miniMap: Maybe<Record<string, any>>) {
  let serialized: string, root: TIsolate;

  withRunTime(() => {
    root = Isolate.create(
      'URoot',
      () => {
        Isolate.create('UChild_1', () => {}, { some_data: true });
        Isolate.create('UChild_2', () => {});
        Isolate.create('UChild_3', () => {});
      },
      {
        some_data: true,
      },
    );

    serialized = IsolateSerializer.serialize(root, miniMap);
  });

  // @ts-ignore
  return { root, serialized };
}
