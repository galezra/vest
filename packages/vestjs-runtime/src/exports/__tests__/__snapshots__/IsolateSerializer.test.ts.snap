// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IsolateSerializer > Custom Data Serialization > Should serialize data with custom keys 1`] = `"[{"0":"URoot","1":{"2":"3"},"children":[{"0":"UChild_1","1":{"2":"3"}},{"0":"UChild_2"},{"0":"UChild_3"}]},{"0":"$type","1":"data","2":"some_data","3":true}]"`;

exports[`IsolateSerializer > serialize > Should produce serialized dump 1`] = `"[{"0":"URoot","1":{"2":"3"},"children":[{"0":"UChild_1","1":{"2":"3"}},{"0":"UChild_2"},{"0":"UChild_3"}]},{"0":"$type","1":"data","2":"some_data","3":true}]"`;
